### PR TITLE
Make transpose optimizer able to look past DQ node for const initializer

### PR DESCRIPTION
**Pull request by skottmckay**
### Description

Add ability for transpose optimizer to look past a DQ node if it has a constant initializer as input. This allows UnsqueezeInput/TransposeInput to modify the initializer in-place in the same way it would for a non-QDQ format model.

Shared initializers are also handled, and any additional Squeeze/Transpose added to the other usages of the initializer should cancel out when we push the same Transpose though them.

The in-place modification means we don't need to run QDQ fixup and constant folding after layout transformation. This means we do not need to enable those optimizers in a minimal build to get an optimal model post-layout transformation.

### Motivation and Context

Ensure layout transformation produces optimal model in full and minimal builds.

### CODE REVIEW

File: onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc [hunk 1]
```diff
@@ -357,51 +414,115 @@ static std::string_view HelpHandleUnsqueeze(HandlerArgs& args, const std::vector
...
 
   // Case 2: input is a Squeeze node with matching axes
   std::unique_ptr<api::NodeRef> inp_node = ctx.graph.GetNodeProducingOutput(input);
+
+  // check if this is a special-cased DQ node where we put the Squeeze prior to it in 'Case 1' above
+  if (inp_node && inp_node->OpType() == "DequantizeLinear" &&
+      std::find_if(ctx.special_cased_dq_nodes.begin(), ctx.special_cased_dq_nodes.end(),
+                   [&inp_node](int64_t id) {
+                     return id == inp_node->Id();
+                   }) != ctx.special_cased_dq_nodes.end()) {
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:483]: Need to use find_if with updated setup as it has to check more details.

File: onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc [hunk 2]
```diff
@@ -345,6 +396,11 @@ static std::vector<int64_t> SortedAxesForTransposedInput(const std::vector<int64
   return new_axes;
 }
 
+static void UpdateDQNodeInputAndShape(api::GraphRef& graph, api::NodeRef& dq, std::string_view new_input) {
+  dq.SetInput(0, new_input);
+  auto new_shape = *graph.GetValueInfo(new_input)->Shape();
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:401]: Is there a concern it wouldn't? This only involves initializers so a shape should always be available.

[the remark this answers is not in the document]
Comment (edgchen1) [reply onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:401]: ah, in that case maybe not. was just looking at it as a helper function.

File: onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc [hunk 3]
```diff
@@ -453,62 +583,153 @@ static void Permute1DConstant(api::GraphRef& graph, api::NodeRef& node, api::Ten
...
+    //   e.g. it provides a set of values that are relative to the input axes like the `sizes` input for Resize
+    // TODO: TransposeInitializer could be updated to handle this case.
     // Permute1DConstant permutes the constant and adds a new initializer. The old initializer is removed only if
     // there are no other consumers.
+    // NOTE: As this returns after calling Permute1DConstant we're implicitly assuming there are no
+    //       other consumers that need updating. This would typically be the case for this sort of input though.
     if (constant->Shape().size() == 1 && constant->Shape()[0] == gsl::narrow_cast<int64_t>(perm.size())) {
-      Permute1DConstant(graph, node, *constant, i, input, perm);
+      assert(consumers->nodes.size() == 0);
+      Permute1DConstant(graph, node, *constant, i, value_to_modify, perm);
```
Comment (adrianlizarraga) [inline onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:635]: Does this have to account for the possibility of a DQ node between the 1D constant and the node? E.g., the `scales` input for Resize could have been originally quantized. This seems very unlikely, but could it be possible? I don't know if we should handle it, assert it, or write a comment.

File: onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc [hunk 4]
```diff
@@ -453,62 +583,153 @@ static void Permute1DConstant(api::GraphRef& graph, api::NodeRef& node, api::Ten
...
+  auto value_to_modify = dq_node ? constant_dq_input : input;
+  auto consumers = graph.GetValueConsumers(value_to_modify);
 
   // Case 1: input is a constant with a known list of consumer nodes
   if (constant != nullptr && consumers->comprehensive) {
-    // Input is scalar, return early.
-    if (constant->Shape().size() == 1 && constant->Shape()[0] == 0) {
+    // If there is only one element return early as the transpose won't change the data
+    if (constant->NumElements() == 1) {
       return;
```
Comment (adrianlizarraga) [inline onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:623]: Just wondering, why is the original input connection not restored before returning? (`node.SetInput(i, input)`)

File: onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc (hunk 3 above)
Comment (skottmckay) [reply onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:635]: AFAIK there's no use-case where we'd quantize the values for things like `scales` in a Resize. Should be relatively simple to pass in a DQ if found though so I'll make it do that.

File: onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc (hunk 4 above)
Comment (skottmckay) [reply onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:639]: Original code was incorrect. Added checks in unit test that things are reconnected.

File: onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc [hunk 5]
```diff
@@ -540,11 +786,25 @@ void TransposeInput(api::GraphRef& graph, api::NodeRef& node, size_t i,
...
 
+void TransposeInput(api::GraphRef& graph, api::NodeRef& node, size_t i,
+                    const std::vector<int64_t>& perm,
+                    const std::vector<int64_t>& perm_inv) {
+  // this TransposeInput is used by the layout transformer to wrap a node in Transpose ops. there's no OptimizerCtx
+  // in that scenario and we're not tracking special-cased DQ nodes as we only do that when pushing Transpose nodes.
+  TransposeInputImpl(graph, /* nodes_using_updated_shared_initializer */ nullptr, node, i, perm, perm_inv);
+}
+
+void TransposeInput(OptimizerCtx& ctx, api::NodeRef& node, size_t i, const std::vector<int64_t>& perm,
```
Comment (adrianlizarraga) [inline onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc:801]: nit: should this be static?

**Review by adrianlizarraga — DISMISSED** [review]

File: onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc (hunk 5 above)
Comment (skottmckay) [reply onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc]: Good catch. Done.

**Review by adrianlizarraga — APPROVED** [review]

**Pull request merged**